orquesta/conducting.py

```diff
@@ -730,6 +761,50 @@ def update_task_state(self, task_id, route, event):
             state_ctx = {'__state': self.workflow_state.serialize()}
             current_ctx = dict_util.merge_dicts(current_ctx, state_ctx, True)
 
+            # This decides whether this task would be retried or not when 'retry' is specified
+            may_retry = (
+                hasattr(task_spec, 'retry') and
+                hasattr(task_spec.retry, 'when') and
+                task_spec.has_retry()
+            )
+            if may_retry:
+                spec_retry_count = expr_base.evaluate(task_spec.retry.count, current_ctx)
+
+                # This is a measure to pass through this processing safely when invalid values
+                # are specified in retry parameter
+                try:
+                    # Current task will be retried when all these conditions are satisfied
+                    #   - new status is same as the one which is specified one.
+                    #   - task has never been repeated for specified times.
+                    will_retry = (
+                        int(spec_retry_count) > task_state_entry['ctxs']['retry']['count'] and
+                        expr_base.evaluate(task_spec.retry.when, current_ctx)
+                    )
+                    if will_retry:
+                        # This turns task_status back to old one.
+                        task_state_entry['status'] = new_task_status = old_task_status
```

Comment:
I am struggling with this. The state machine `machines.TaskStateMachine.process_event` at line https://github.com/StackStorm/orquesta/blob/master/orquesta/conducting.py#L706 should be the one to determine the task status.  This is basically going outside of the state machine to make changes to the task status. Is it possible to let the TaskStateMachine set the new task status to `retrying` and the `if may_retry:` above is just `if new_task_status == statuses.RETRYING:`?

Reply:
I wanted to process task-retry by using [the state transition machine for task](https://github.com/StackStorm/orquesta/blob/master/orquesta/machines.py#L551). But it's difficult to implement because retrying processing is non deterministic. Therefore, I implemented it to rollback task status by using evaluator.

---

What makes it difficult to do it is state transition is determined by following three factors.
  - current task's status
  - current taks's retry-count
  - user defined condition which is specified in 'when' parameter in metadata

When I define task transitions for each status to take into account the retry-count and user-specified conditions, it might be possible to implement this state transition processing by using TASK_STATE_MACHINE_DATA.

I also believe that it's more elegant implementation. But It's hard for me to implement it without omission and bugs. And it's easy to implement the processing to determine whether task should be retried by using evaluator. That takes into account both current status and user specified condition to retry at the same time.

So please forgive me to implement that feature to rollback status after changing status when state-machine determines task will be retried.